Excess properties not detected when assigning to union type containing a function type

The code below should throw a compiler error or am i missing something?

Code on Playground: [https://www.typescriptlang.org/play](https://www.typescriptlang.org/play/#src=type%20NormalType%20%3D%20string%3B%0D%0Atype%20WorksAsExpected%20%3D%20{a%3Anumber%2C%20c%3Anumber}%20|%20NormalType%3B%0D%0A%0D%0Alet%20works%3AWorksAsExpected%20%3D%20{a%3A1%2C%20c%3A2%2C%20d%3A3}%20%2F%2Ferror%2C%20works%0D%0A%0D%0Atype%20FunctionType%20%3D%20%28%29%3D%3Eany%3B%0D%0Atype%20DoesntWork%20%3D%20{a%3Anumber%2C%20c%3Anumber}%20|%20FunctionType%3B%0D%0A%0D%0Alet%20worksTo%3ADoesntWork%20%3D%203%20%2F%2Ferror%2C%20works%20too%0D%0Alet%20doesntWork%3ADoesntWork%20%3D%20{a%3A1%2C%20c%3A2%2C%20d%3A3}%20%2F%2Ferror%2C%20doesn%27t%20work)

**TypeScript Version:**  
typescript@1.8.10, typescript@1.9.0-dev.20160519-1.0

**Code**

```ts
type NormalType = string;
type WorksAsExpected = {a:number, c:number} | NormalType;

let works:WorksAsExpected = {a:1, c:2, d:3} 
/*erro.ts(4,40): error TS2322: Type '{ a: number; c: number; d: number; }' 
is not assignable to type '{ a: number; c: number; } | string'.
  Object literal may only specify known properties, and 'd' does not exist in type 
'{ a: number; c: number; } | string'.*/

type FunctionType = ()=>any;
type DoesntWork = {a:number, c:number} | FunctionType;

let worksTo:DoesntWork = 3 
/*erro.ts(9,5): error TS2322: Type 'number' is not assignable to type 
'{ a: number; c: number; } | (() => any)'.
  Type 'number' is not assignable to type '() => any'.
*/

let doesntWork:DoesntWork = {a:1, c:2, d:3} //no error is displayed

```

**Expected behavior:**
The third assignment should cause a compiler error
**Actual behavior:** 
No error is displayed

--- a/src/compiler/checker.ts
+++ b/src/compiler/checker.ts
@@ -6030,7 +6030,7 @@ namespace ts {
                 if (type.flags & TypeFlags.ObjectType) {
                     const resolved = resolveStructuredTypeMembers(type);
                     if ((relation === assignableRelation || relation === comparableRelation) &&
-                        (type === globalObjectType || resolved.properties.length === 0) ||
+                        (type === globalObjectType || isEmptyObjectType(resolved)) ||
                         resolved.stringIndexInfo || resolved.numberIndexInfo || getPropertyOfType(type, name)) {
                         return true;
                     }
@@ -6043,6 +6043,14 @@ namespace ts {
                     }
                 }
                 return false;
+            }
+
+            function isEmptyObjectType(t: ResolvedType) {
+                return t.properties.length === 0 &&
+                    t.callSignatures.length === 0 &&
+                    t.constructSignatures.length === 0 &&
+                    !t.stringIndexInfo &&
+                    !t.numberIndexInfo;
             }
 
             function hasExcessProperties(source: FreshObjectLiteralType, target: Type, reportErrors: boolean): boolean {

--- a/tests/baselines/reference/excessPropertyErrorForFunctionTypes.errors.txt
+++ b/tests/baselines/reference/excessPropertyErrorForFunctionTypes.errors.txt
@@ -1,0 +1,12 @@
+tests/cases/compiler/excessPropertyErrorForFunctionTypes.ts(4,44): error TS2322: Type '{ a: number; c: number; d: number; }' is not assignable to type '{ a: number; c: number; } | (() => any)'.
+  Object literal may only specify known properties, and 'd' does not exist in type '{ a: number; c: number; } | (() => any)'.
+
+
+==== tests/cases/compiler/excessPropertyErrorForFunctionTypes.ts (1 errors) ====
+    type FunctionType = () => any;
+    type DoesntWork = { a: number, c: number } | FunctionType;
+    
+    let doesntWork: DoesntWork = { a: 1, c: 2, d: 3 }
+                                               ~~~~
+!!! error TS2322: Type '{ a: number; c: number; d: number; }' is not assignable to type '{ a: number; c: number; } | (() => any)'.
+!!! error TS2322:   Object literal may only specify known properties, and 'd' does not exist in type '{ a: number; c: number; } | (() => any)'.

--- a/tests/baselines/reference/excessPropertyErrorForFunctionTypes.js
+++ b/tests/baselines/reference/excessPropertyErrorForFunctionTypes.js
@@ -1,0 +1,8 @@
+//// [excessPropertyErrorForFunctionTypes.ts]
+type FunctionType = () => any;
+type DoesntWork = { a: number, c: number } | FunctionType;
+
+let doesntWork: DoesntWork = { a: 1, c: 2, d: 3 }
+
+//// [excessPropertyErrorForFunctionTypes.js]
+var doesntWork = { a: 1, c: 2, d: 3 };

--- a/tests/cases/compiler/excessPropertyErrorForFunctionTypes.ts
+++ b/tests/cases/compiler/excessPropertyErrorForFunctionTypes.ts
@@ -1,0 +1,4 @@
+type FunctionType = () => any;
+type DoesntWork = { a: number, c: number } | FunctionType;
+
+let doesntWork: DoesntWork = { a: 1, c: 2, d: 3 }

